### PR TITLE
Migrate to use com.vanniktech.maven.publish

### PR DESCRIPTION
**Pull request by pkwarren**
Update the buf-gradle-plugin build to use the
`com.vanniktech.maven.publish` plugin instead of the `io.github.gradle-nexus.publish-plugin`. Move all of the build configuration into `build.gradle.kts` and update CI workflows to migrate off deprecated GitHub actions and action inputs.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -21,17 +21,17 @@ jobs:
         with:
           fetch-depth: 0
       - name: Validate Gradle wrapper
-        uses: gradle/wrapper-validation-action@v3
+        uses: gradle/actions/wrapper-validation@v3
```
Comment (pkwarren) [inline .github/workflows/ci.yaml:24]: https://github.com/gradle/actions/blob/main/docs/deprecation-upgrade-guide.md#the-action-gradlewrapper-validation-action-has-been-replaced-by-gradleactionswrapper-validation

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -21,17 +21,17 @@ jobs:
...
         with:
           distribution: temurin
           java-version: 11
           check-latest: true
+      - name: Setup Gradle
+        uses: gradle/actions/setup-gradle@v3
       - name: Build
-        uses: gradle/gradle-build-action@v3
-        with:
-          arguments: build --info
```
Comment (pkwarren) [inline .github/workflows/ci.yaml:34]: https://github.com/gradle/actions/blob/main/docs/deprecation-upgrade-guide.md#using-the-action-to-execute-gradle-via-the-arguments-parameter-is-deprecated

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -40,14 +40,14 @@ jobs:
           comment_mode: off
           check_name: "Test Results (Ubuntu)"
       - name: Publish unit test results (Windows)
-        uses: EnricoMi/publish-unit-test-result-action/composite@v2
+        uses: EnricoMi/publish-unit-test-result-action/windows@v2
```
Comment (pkwarren) [inline .github/workflows/ci.yaml:43]: https://github.com/EnricoMi/publish-unit-test-result-action/tree/v2?tab=readme-ov-file#running-as-a-composite-action

File: .github/workflows/release.yaml [hunk 4]
```diff
@@ -16,27 +16,27 @@ jobs:
...
-            closeAndReleaseSonatypeStagingRepository
-            publishPlugins
-            -Dorg.gradle.internal.http.socketTimeout=120000
-            -Dorg.gradle.internal.network.retry.max.attempts=1
-            -Dorg.gradle.internal.publish.checksums.insecure=true
+        run: |
+          ./gradlew
+          publishToMavenCentral
+          publishPlugins
+          --no-configuration-cache
```
Comment (pkwarren) [inline .github/workflows/release.yaml:33]: https://vanniktech.github.io/gradle-maven-publish-plugin/central/#automatic-release

File: .github/workflows/release.yaml [hunk 5]
```diff
@@ -16,27 +16,27 @@ jobs:
...
+          -Dorg.gradle.internal.network.retry.max.attempts=1
         env:
-          OSSRH_USERNAME: ${{ secrets.SONATYPE_USER }}
-          OSSRH_PASSWORD: ${{ secrets.SONATYPE_PASSWORD }}
-          PGP_KEY: ${{ secrets.GPG_KEY_ARMORED }}
-          PGP_PASSWORD: ${{ secrets.GPG_PASSPHRASE }}
+          ORG_GRADLE_PROJECT_mavenCentralUsername: ${{ secrets.SONATYPE_USER }}
+          ORG_GRADLE_PROJECT_mavenCentralPassword: ${{ secrets.SONATYPE_PASSWORD }}
+          ORG_GRADLE_PROJECT_signingInMemoryKey: ${{ secrets.GPG_KEY_ARMORED }}
+          ORG_GRADLE_PROJECT_signingInMemoryKeyPassword: ${{ secrets.GPG_PASSPHRASE }}
```
Comment (pkwarren) [inline .github/workflows/release.yaml:40]: https://vanniktech.github.io/gradle-maven-publish-plugin/central/#__tabbed_5_2

File: build.gradle.kts [hunk 6]
```diff
@@ -61,7 +69,42 @@ kotlin {
     }
 }
 
+mavenPublishing {
+    publishToMavenCentral(SonatypeHost.S01)
```
Comment (pkwarren) [inline build.gradle.kts:73]: @rubensf - This is what we would change if we wanted to publish to central instead of s01.

**Review by rubensf — APPROVED** [review]

**Pull request merged**